File: ai-research-copilot/backend/app/schemas/document.py
"""
Document Pydantic v2 schemas.

Request/response schemas for document CRUD, upload, chunking, and listing.
"""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class DocumentCreate(BaseModel):
    """Schema for creating a new document."""

    model_config = ConfigDict(str_strip_whitespace=True)

    name: Optional[str] = Field(
        default=None,
        max_length=255,
        description="Human-readable document name. Defaults to the original filename if omitted.",
    )
    knowledge_base_ids: Optional[list[UUID]] = Field(
        default=None,
        description="Knowledge base IDs to associate this document with on creation.",
    )


class DocumentUpdate(BaseModel):
    """Schema for updating an existing document."""

    model_config = ConfigDict(str_strip_whitespace=True)

    name: Optional[str] = Field(
        default=None,
        max_length=255,
        description="Updated human-readable document name.",
    )


class DocumentResponse(BaseModel):
    """Schema for document data returned in API responses."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    name: str
    original_filename: str
    mime_type: str
    file_size: int
    status: str
    chunk_count: int
    created_at: datetime
    updated_at: datetime


class DocumentDetail(DocumentResponse):
    """Extended document schema with storage and processing details."""

    storage_backend: str
    content_text: Optional[str] = None
    processing_error: Optional[str] = None
    metadata_extra: Optional[dict] = None


class DocumentList(BaseModel):
    """Paginated list of documents."""

    model_config = ConfigDict(from_attributes=True)

    items: list[DocumentResponse]
    total: int
    page: int
    page_size: int


class DocumentUploadResponse(BaseModel):
    """Response returned after a document upload is accepted."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    status: str
    message: str


class DocumentChunkResponse(BaseModel):
    """Schema for a single document chunk."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    chunk_index: int
    content: str
    token_count: int
    chunk_type: str
    metadata_extra: Optional[dict] = None


class DocumentChunkList(BaseModel):
    """Paginated list of document chunks."""

    model_config = ConfigDict(from_attributes=True)

    items: list[DocumentChunkResponse]
    total: int

File: ai-research-copilot/frontend/src/types/document.ts
export interface Document {
  id: string;
  name: string;
  original_filename: string;
  mime_type: string;
  file_size: number;
  status: "pending" | "processing" | "completed" | "failed";
  chunk_count: number;
  embedding_status: "pending" | "processing" | "completed" | "failed";
  error_message: string | null;
  metadata: Record<string, unknown> | null;
  knowledge_base_ids: string[];
  created_at: string;
  updated_at: string;
}

export interface DocumentDetail extends Document {
  chunks: DocumentChunk[];
}

export interface DocumentChunk {
  id: string;
  document_id: string;
  content: string;
  chunk_index: number;
  token_count: number;
  metadata: Record<string, unknown> | null;
  created_at: string;
}

export interface DocumentUploadResponse {
  id: string;
  name: string;
  status: string;
  message: string;
}

export interface DocumentList {
  items: Document[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface DocumentState {
  documents: Document[];
  currentDocument: DocumentDetail | null;
  uploadQueue: UploadItem[];
  isLoading: boolean;
  error: string | null;
  setDocuments: (documents: Document[]) => void;
  setCurrentDocument: (doc: DocumentDetail | null) => void;
  addToUploadQueue: (item: UploadItem) => void;
  updateUploadProgress: (id: string, progress: number) => void;
  updateUploadStatus: (
    id: string,
    status: UploadItem["status"],
    error?: string,
    documentId?: string
  ) => void;
  removeFromUploadQueue: (id: string) => void;
  clearUploadQueue: () => void;
  setIsLoading: (isLoading: boolean) => void;
  setError: (error: string | null) => void;
}

export interface UploadItem {
  id: string;
  file: File;
  progress: number;
  status: "pending" | "uploading" | "completed" | "error";
  error?: string;
  documentId?: string;
}
